Restore:  improve DependencyGraphSpec memory usage

dtivel opened the pull request:
## Bug

Fixes:  https://github.com/NuGet/Home/issues/9040 
Regression:  No

## Fix

Details:
1. Obsolesce the unnecessary `DependencyGraphSpec.Json` property.
2. Load/read a dependency graph spec file using a forward only JSON reader.

## Testing/Validation

Tests Added:  Yes 
Validation:  

Measuring the improvement in Visual Studio is difficult because there is too much noise to get a clean comparison.  To isolate the impact of this change, I created a simple console app that loaded all 84 NuGet Client project .dgspec.json files using the old and new readers:
```C#
// My VM had 4 processors
var options = new ParallelOptions() { MaxDegreeOfParallelism = Environment.ProcessorCount };

// Fire an ETW start event
eventSource.LoadAllStart();

try
{
    Parallel.ForEach(DgSpecFilePaths, options, dgSpecFilePath => DependencyGraphSpec.Load(dgSpecFilePath));
}
finally
{
    eventSource.LoadAllStop();
}
```
Based on 5 iterations each (before and after), the average improvements are:
| Metric | Before | After | Delta | % Delta |
| --- | --- | --- | --- | --- |
| total time reading all files in parallel | 385.4 ms | 289.2 ms | -96.2 ms | -25.0% |
| heap allocations | 58.8 MB | 33.2 MB | -25.6 MB | -43.5% |

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -247,6 +237,74 @@ public void Save(string path)
             }
         }
 
+        internal static DependencyGraphSpec Load(string path, bool useImprovedReader)
```
The only place `useImprovedReader` is set to false is in tests. It was a good way to check for regressions during development, but since the production code only ever uses `true`, should it be deleted now? the `false` codepath is now dead code.

cristinamanum commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -75,11 +72,6 @@ public IReadOnlyList<PackageSpec> Projects
-        /// <summary>
-        /// File json.
-        /// </summary>
-        public JObject Json { get; }
```
This is public, can it be removed? Maybe obsolete it?

cristinamanum commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -254,9 +246,61 @@ public static DependencyGraphSpec Union(IEnumerable<DependencyGraphSpec> dgSpecs
+                        switch (propertyName)
+                        {
+                            case "format":
+                                format = jsonReader.ReadAsInt32();
```
Is the format used anywhere?

dtivel replied:
Ha ha, no.  It wasn't even read in the old reader.  Silly that we created a format property and never read it.  I'll remove this.

cristinamanum commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -917,49 +917,50 @@ private static void PopulateDownloadDependencies(IList<DownloadDependency> downl
+        private static void PopulateCentralPackageVersions(
            string packageSpecPath,
-           TargetFrameworkInformation targetFrameworkInformation,
+           IDictionary<string, CentralPackageVersion> centralPackageVersons,
```
centralPackageVersons -> centralPackageVersions

cristinamanum commented on src/NuGet.Core/NuGet.ProjectModel/JsonTextReaderExtensions.cs:
```diff
@@ -0,0 +1,208 @@
+                    case JsonToken.String:
+                        var value = (string)reader.Value;
+
+                        return value.Split(new[] { ' ', ',' }, StringSplitOptions.RemoveEmptyEntries);
```
Would it be worth to add the delimiters as params argument?

dtivel replied:
No, not now.  If there's a compelling use case, I'd consider it.

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -78,7 +78,8 @@ public IReadOnlyList<PackageSpec> Projects
         /// </summary>
-        public JObject Json { get; }
+        [Obsolete("This property is obsolete and will be removed in a future release.")]
+        public JObject Json { get => _json.Value; }
```
This feels off. I get that it's not gonna be a missing method exception like it would've been if we had removed it, but returning an empty JObject is weird and likely to break anyone using it in a similar way.

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/JsonTextReaderExtensions.cs:
```diff
@@ -0,0 +1,210 @@
+{
+    internal static class JsonTextReaderExtensions
+    {
+        private static readonly char[] DelimitedStringDelimiters = new char[] { ' ', ',' };
```
Where do we use ` ` (space) as a delimeter?

dtivel replied:
[Here](https://github.com/NuGet/NuGet.Client/blob/4c171de8f66650b7782eceaa8b0feedf6c8a9f6c/src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs#L663).  Any property that uses `TryGetStringEnumerable(…)`.  That includes dependency `type`; files' `include`, `exclude`, and `suppressParent`; and `privateAssets`.

[`NuGet.ProjectModel.Test.IncludeFlagTests.IncludeFlag_RoundTrip(…)`](https://github.com/NuGet/NuGet.Client/blob/4c171de8f66650b7782eceaa8b0feedf6c8a9f6c/test/NuGet.Core.Tests/NuGet.ProjectModel.Test/IncludeFlagTests.cs#L451) is an existing test which covers this case.

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/MemoryEfficientJsonPackageSpecReader.cs:
```diff
@@ -0,0 +1,1570 @@
+            string title = null;
+            string version = null;
+            var wasPackOptionsSet = false;
+            var isMappingsNull = false;
```
Can we have a positive condition name?

dtivel replied:
It _is_ a positive condition name.  `isMappingsNotNull` would be a negative condition not a positive one.  A double negative (e.g.:  `if (!isMappingsNotNull) ...`) [has been avoided](https://github.com/NuGet/NuGet.Client/pull/3239/files#diff-639b72962d98b9eabacc47f0325923acR164).

I get the sense that you feel that "null" is an inherently negative name, but I don't think it is.  It can be part of a negative expression though:  `isNull` or `isNotNull`.  In fact, [`IsNull` is used commonly throughout the .NET Framework](https://referencesource.microsoft.com/#q=IsNull).

What I'm trying to do here is reproduce the same error condition that `JsonPackageSpecReader` produces when `packOptions`/`files`/`mappings` property exists but its value is not a valid object so `Mappings` should be set to `null`.

nkolev92 replied:
Makes sense.

nkolev92 reviewed (commented):
I love this change overall and from the first review I don't think there's any discrepancies here, but in general it feels like a lot of duplication. 

A few questions/ideas: 

* Why not do the implementation on the JsonPackageSpecReader side? Do we have a deprecation/full migration plan there?
* Even if we don't go that way, can we extract utilities for the common functionalities? 
* I see 8 different uses of `JsonPackageSpecReader.GetPackageSpec(string, string)` that probably should go through the efficient path. 
* Should we consider doing the same for the assets file reading? Currently it's doing the same JsonUtility.Load.

dtivel committed:
Improve dependency graph spec load performance.

dtivel commented:
@nkolev92, I replaced the implementation of `JsonPackageSpecReader` and deleted `MemoryEfficientJsonPackageSpecReader`.  Also, I added the `Obsolete` attribute to members which use JLinq types in preparation for https://github.com/NuGet/Home/issues/9197.

rrelyea commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -78,7 +84,8 @@ public IReadOnlyList<PackageSpec> Projects
         /// </summary>
-        public JObject Json { get; }
+        [Obsolete("This property is obsolete and will be removed in a future release.")]
+        public JObject Json { get => _json.Value; }
```
should this property also be marked obsolete?

rrelyea commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -47,126 +49,180 @@ public static PackageSpec GetPackageSpec(string json, string name, string packag
+            packageSpec.RuntimeGraph = new RuntimeGraph(
+                runtimeDescriptions ?? Enumerable.Empty<RuntimeDescription>(),
+                compatibilityProfiles ?? Enumerable.Empty<CompatibilityProfile>());
+            packageSpec.Title = title;
```
in the switch statement above...often you set state directly on the packageSpec. Why isn't title done similarly?

rrelyea commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -47,126 +49,180 @@ public static PackageSpec GetPackageSpec(string json, string name, string packag
-            var packInclude = rawPackageSpec["packInclude"] as JObject;
-            if (packInclude != null)
+            packageSpec.Authors = authors ?? Array.Empty<string>();
+            packageSpec.BuildOptions = buildOptions;
```
same question as I have for title below.

rrelyea commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -47,126 +49,180 @@ public static PackageSpec GetPackageSpec(string json, string name, string packag
-                {
-                    packageSpec.PackInclude.Add(new KeyValuePair<string, string>(include.Key, include.Value.ToString()));
-                }
+                packageSpec.ContentFiles = contentFiles;
```
same question as I have for title below.

rrelyea commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -618,445 +501,1122 @@ private static PackOptions GetPackOptions(PackageSpec packageSpec, JObject rawPa
+                        string projectStyleString = jsonReader.ReadNextTokenAsString();
+
+                        if (!string.IsNullOrEmpty(projectStyleString)
+                            && Enum.TryParse<ProjectStyle>(projectStyleString, ignoreCase: true, result: out var projectStyleValue))
```
nit: don't think you need the var here...is there a reason you include?

donnie-msft replied:
I think he wants to preserve `projectStyle` in case the Parsing fails. If he doesn't create a new variable, then it'll overwrite it with null upon failure.

dtivel replied:
@donnie-msft is correct.

Like it or not, that was the [existing behavior](https://github.com/NuGet/NuGet.Client/blob/8a1d1c01bc27c282d20668dbbe1abd2182f9b3fb/src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs#L208-L213).  I only preserve it.

rrelyea commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -618,445 +501,1122 @@ private static PackOptions GetPackOptions(PackageSpec packageSpec, JObject rawPa
+                msbuildMetadata.RestoreLockProperties = restoreLockProperties;
+            }
+
+            msbuildMetadata.SkipContentFileWrite = skipContentFileWrite;
```
Seems like having the parsing and the setting of the data seperate, is somewhat risky... what if you harvest data, but forget to set it.  I would lean towards setting the value while harvesting, when possible...and having a minimal amount of state set after the harvesting operations are all done.

dtivel replied:
For several of the places you pointed this out elsewhere, I've fixed them.  However, for restore metadata this unavoidable:  we have to read the project style before we can even create `msbuildMetadata`.  It could be either an instance of `PackagesConfigProjectRestoreMetadata` or `ProjectRestoreMetadata` --- we don't know until we've read all properties.

rrelyea commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -618,445 +501,1122 @@ private static PackOptions GetPackOptions(PackageSpec packageSpec, JObject rawPa
+                        break;
+
+                    case "packageType":
+                        ReadPackageTypes(packageSpec, jsonReader);
```
ReadPackageTypes() method is different style than all your other Read*() methods. Should your naming pattern differ? (you pass in the packageSpec in this case)

dtivel replied:
Actually, a lot of Read*() methods pass `PackageSpec` instance.  The reason for this is entirely for performance reasons.  `PackageSpec` has many properties which are initialized to an empty collection, and also many properties which are `null` by default.  When the properties are already initialized to a collection, I'd rather use that collection instead of creating _another_ collection and overwriting the `PackageSpec` instance's collection.  The latter practice is what was happening all over the place before this change.

What naming pattern do you suggest?

rrelyea commented on test/NuGet.Core.Tests/NuGet.ProjectModel.Test/ImportFrameworkTest.cs:
```diff
@@ -49,8 +48,8 @@ public void ImportFramwork_EmptyImport()
-            var spec = JsonPackageSpecReader.GetPackageSpec(configJson.ToString(), "TestProject", "project.json");
-            var importFramework = spec.TargetFrameworks.First().Imports.ToList();
+            PackageSpec spec = JsonPackageSpecReader.GetPackageSpec(configJson.ToString(), "TestProject", "project.json");
+            System.Collections.Generic.List<NuGetFramework> importFramework = spec.TargetFrameworks.First().Imports.ToList();
```
nit: using System.Collections.Generic?

nkolev92 reviewed (approved):
👏

dtivel committed:
Apply feedback.
